xds/internal/xdsclient/tests/ads_stream_flow_control_test.go

```diff
@@ -0,0 +1,623 @@
+//     listener watchers for the two resources, and verifies that the update
+//     reaches these watchers.
+//   - These watchers don't invoke the onDone callback until explicitly
+//     triggered by the test. This allows the test to verify that the next
+//     Recv() call on the ADS stream does not happen until both watchers have
+//     completely processed the update, i.e invoke the onDone callback.
+func (s) TestADSFlowControl_ResourceUpdates_MultipleResources(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	// Override the ADS stream creation.
+	adsStreamCh := overrideADSStreamCreation(t)
+
+	// Start an xDS management server.
+	const listenerResourceName1 = "test-listener-resource-1"
+	const listenerResourceName2 = "test-listener-resource-2"
+	wantResourceNames := []string{listenerResourceName1, listenerResourceName2}
+	requestCh := make(chan struct{})
+	mgmtServer := e2e.StartManagementServer(t, e2e.ManagementServerOptions{
+		OnStreamRequest: func(id int64, req *v3discoverypb.DiscoveryRequest) error {
+			if req.GetTypeUrl() != version.V3ListenerURL {
+				return nil
+			}
+			gotResourceNames := req.GetResourceNames()
+			sort.Slice(gotResourceNames, func(i, j int) bool { return req.ResourceNames[i] < req.ResourceNames[j] })
+			if slices.Equal(gotResourceNames, wantResourceNames) {
+				// The two resource names will be part of the initial request
+				// and also the ACK. Hence, we need to make this write
+				// non-blocking.
+				select {
+				case requestCh <- struct{}{}:
+				default:
+				}
+			}
+			return nil
+		},
+	})
+
+	// Create bootstrap configuration pointing to the above management server.
+	nodeID := uuid.New().String()
+	bc := e2e.DefaultBootstrapContents(t, nodeID, mgmtServer.Address)
+	testutils.CreateBootstrapFileForTesting(t, bc)
+
+	// Create an xDS client with the above bootstrap contents.
+	client := createXDSClient(t, bc)
+
+	// Configure two watchers for two different listener resources.
+	const routeConfigurationName1 = "test-route-configuration-resource-1"
+	watcher1 := newBLockingListenerWatcher()
+	cancel1 := xdsresource.WatchListener(client, listenerResourceName1, watcher1)
+	defer cancel1()
+	const routeConfigurationName2 = "test-route-configuration-resource-2"
+	watcher2 := newBLockingListenerWatcher()
+	cancel2 := xdsresource.WatchListener(client, listenerResourceName2, watcher2)
+	defer cancel2()
+
+	// Wait for the wrapped ADS stream to be created.
+	var adsStream *wrappedADSStream
+	select {
+	case adsStream = <-adsStreamCh:
+	case <-ctx.Done():
+		t.Fatalf("Timed out waiting for ADS stream to be created")
+	}
+
+	// Ensure that there is a read on the stream.
+	select {
+	case <-adsStream.recvCh:
+	case <-ctx.Done():
+		t.Fatalf("Timed out waiting for ADS stream to be read from")
+	}
+
+	// Wait for both resource names to be requested.
+	select {
+	case <-requestCh:
+	case <-ctx.Done():
```

Comment:
Isn't this flaky? What if the code in the onStreamRequest function finishes processing, there's no corresponding recv here since doesn't get to this part of test (this didn't apply to my comment above since those channels are buffered).

Reply:
That's a fair point. Changed the channel to have a buffer of size `1`. I usually do that, I don't know how/why I missed it here. Thanks for pointing out.